File: 0486-predict-the-winner/0486-predict-the-winner.ts
function predictTheWinner(nums: number[]): boolean {
    const n = nums.length;
    if ((n & 1) === 0) return true; // even length

    const dp: number[] = new Array(n);

    for (let i = n - 1; i >= 0; i--) {
        dp[i] = nums[i];
        
        for (let j = i + 1; j < n; j++) {
            dp[j] = Math.max(nums[i] - dp[j], nums[j] - dp[j - 1]);
        }
    }
    return dp[n - 1] >= 0;
}

File: 0486-predict-the-winner/0486-predict-the-winner.rs
impl Solution {
    pub fn predict_the_winner(nums: Vec<i32>) -> bool {
        let n = nums.len();
        if n % 2 == 0 {
            return true;
        }

        let mut dp = vec![0; n];
        for i in (0..n).rev() {
            dp[i] = nums[i];
            for j in i + 1..n {
                dp[j] = (nums[i] - dp[j]).max(nums[j] - dp[j - 1]);
            }
        }
        dp[n - 1] >= 0
    }
}